resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/SseEventOutputImpl.java

```diff
@@ -85,8 +85,17 @@ protected synchronized void flushResponseToClient()
    {
       if (!responseFlushed)
       {
-         //set back to client 200 OK to implies the SseEventOutput is ready
-         BuiltResponse jaxrsResponse = (BuiltResponse) Response.ok().type(MediaType.SERVER_SENT_EVENTS).build();
+         BuiltResponse jaxrsResponse = null;
+		 if(this.closed)
+		 {
+			 jaxrsResponse = (BuiltResponse) Response.noContent().build();
+		 }
+		 else
+		 {
+			//set back to client 200 OK to implies the SseEventOutput is ready
+			 jaxrsResponse = (BuiltResponse) Response.ok().type(MediaType.SERVER_SENT_EVENTS).build();
+		 }
+		 
```

Comment:
With the change I did in close() method (https://github.com/resteasy/Resteasy/pull/1361/files/2edd2389102cbfc199d881de0dadff73707b3071#diff-97c04cdfc0891675b26b1a572dba739cR78) calling `SseEventSink.close()` will only complete the asynchContext. Method flushResponseToClient() will be invoked when resource method returns since it was added to `Cleanables` by SseEventSinkInterceptor (https://github.com/resteasy/Resteasy/blob/master/resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/providers/sse/SseEventSinkInterceptor.java#L58)

Reply:
Thanks. This looks good.

Reply:
Hey @jimma two more points about this class:

- invocation of **SseEventOutputImpl#flushResponseToClient()** at line 128 (https://github.com/resteasy/Resteasy/pull/1361/files#diff-97c04cdfc0891675b26b1a572dba739cL128) should be done in the try/catch block in order to follow the doc requirement about sending event:

> If there is a problem during sending of an event, completion stage will be completed exceptionally.

- **SseEventOutputImpl#send(OutboundSseEvent event, BiConsumer<SseEventSink, Throwable> errorConsumer)** should be marked as **synchronized** to prevent following race condition:

Thread 1 calls **SseEventOutputImpl#send(OutboundSseEvent event, BiConsumer<SseEventSink, Throwable> errorConsumer)** and since it is not closed  goes to line 131 (https://github.com/resteasy/Resteasy/pull/1361/files#diff-97c04cdfc0891675b26b1a572dba739cR137) without entering the method.
Thread 2 calls **SseEventOutputImpl#close()** so the outputstream is being closed. Then thread 1 continues entering **SseEventOutputImpl#writeEvent(OutboundSseEvent event)** and since the outputstream is closed it will fail with a IOException being thrown and **SseEventOutputImpl#close()** being called for nothing a second time.

Actually it's not really bad since object internal state is still consitent but just a kind of incoherent. Making the method **synchronized** will prevent this behavior without really adding more blocking since both **flushResponseToClient()** and **writeEvent(...)** are already **synchronized**.

What do you think ?

Reply:
>invocation of SseEventOutputImpl#flushResponseToClient() at line 128 (https://github.com/resteasy/Resteasy/pull/1361/files#diff-97c04cdfc0891675b26b1a572dba739cL128) should be done in the try/catch block in order to follow the doc requirement about sending event:
+1.  This requires a try/catch block.
> SseEventOutputImpl#send(OutboundSseEvent event, BiConsumer<SseEventSink, Throwable> errorConsumer) should be marked as synchronized to prevent following race condition
I think close() is the priority operation and it shouldn't wait the send event is done then execute. close() can interrupt a send event operation at any time.

Reply:
Ok so if the close() is the priority let's remove the **synchronized** on it so that it truly does not wait for the send event to be done.
However let's replace the **boolean closed** by an AtomicBoolean to follow spec and be sure that close() will only be invoked once:

>Subsequent calls have no effect and are ignored.

so close() will be like:

	@Override
	public void close() {
		if (!closed.compareAndSet(false, true)) {
			return;
		}
		if (asyncContext.isSuspended() && asyncContext.getAsyncResponse() != null) {
			if (asyncContext.isSuspended()) {
				asyncContext.getAsyncResponse().complete();
			}
		}
	}

Reply:
Hi @jimma,

Forget about the last comment about  thread safety for now please. I will open a new PR to talk about this.
So I've just committed the last fix about the  try/catch block.